### DISCUSSION TITLE
Template API: Should order be a required property for all blocks?

### DISCUSSION THREAD

**Discussion by mattsherman** [woocommerce/woocommerce#39969] [WooCommerce New Product Editor] [answered]
## Background

When we initially implemented the first iteration of `SimpleProductTemplate` using the template API (see https://github.com/woocommerce/woocommerce/pull/39814), we left the `order` property off of all the blocks in the template, inadvertently.

We has since corrected that (see https://github.com/woocommerce/woocommerce/pull/39946).

## Issue

Leaving the `order` property off of the blocks exposed an issue... if there is no explicit `order` defined for blocks, you cannot position blocks between them, as they all have an implicit order value of `10`.

Example:

* Section 1
    * Block A (no order set; implicitly set to `10`)
    * Block B (no order set; implicitly set to `10`)
    * Block C (no order set; implicitly set to `10`)

If you use the template API in an extension to attempt to insert a block in-between "Block A" and "Block C", you'll find that you are unable to, as what would you set the `order` of your block to?

## Proposal

We should make the `order` property a required property for every block added to a template. If the `order` property is not explicitly provided when adding a block, the block will not be added to the template.

I considered having the `order` property automatically increment based on the last block in the container, but I think that is not ideal because:
* The `order` property for such blocks would not be stable, as if other blocks were added in the future before them, the `order` would change
* You can't look at the source code for the template and easily determine what the `order` will be set to.

**Comment by nerrad** [comment]
You've unfortunately revealed a bit of a flaw with using order to control placement. I'm assuming you are referring to the order for _core-provided blocks. It's pretty fragile when extensions start using this to try and refine how their UI is viewed in the context of other extensions.

**Reply by mattsherman** [reply]
Yes, order can definitely be fragile. Early on we debated whether to go with order or whether to go with an API that provided the ability to place blocks in relation to other blocks...

```php
$template->insert_block_after(
    'OTHER_BLOCK_ID', 
    [
        'blockName' => 'BLOCK_NAME',
        ...
    ]
);
```

A downside of that approach is that it just doesn't work at all if the block you are positioning your block in relation to doesn't exist in the template (maybe it was removed by another extension).

A pro of the `order` approach is that it is a bit more resilient to handling that case, since the positioning does not refer to other blocks explicitly. Of course, if you want to really make sure that your block is positioned in relation to another block, you can look up it's `order` at runtime and then base your block's `order` on it. If you do that, you need to be robust in your implementation to handle the case where the other block doesn't exist.

In either approach, you can't guarantee the exact placement of your block, because there can be conflicts between extensions. 

The dependency of blocks on other blocks is an area we are continuing to look into in the API.

**Comment by louwie17** [comment]
Order should either be required or just default to a high number like 10 or a 100, I would lean towards the `default`.

Otherwise I think it should be left up to the template owner how they want to handle order, by first come first serve only using the `default` or being explicit about the order and separating fields by an order of 5 or 10, similar to what you did in your PR above.

Another thought is, maybe we can differentiate between what are the core fields for the template and what fields have been added to it after initialization.
In that case we could add an option to the template generation to auto separate fields by an order of (number of choice) upon initialization.
This doesn't solve the problem of users not being able to figure out the exact order they need, but to solve this we could add a `wp` script that will output the fields with the assigned order for a specific template (until we have come up with another solution to view this in HTML ).

**Reply by joshuatf** [reply, marked as answer]
I'm also tempted to lean towards a default as opposed to requiring order.  That said, a number like `10` or `100` could pretty easily be reached if we add fields in increments of 10s.  For example, if we have a section that has 12 fields, starting at 10, we've already exceeded 100.  If the default order is `100` for blocks, they will be added somewhere before the last couple of items which seems unexpected.

My expectation is that any blocks added without an order should appear at the end of the list.  I think if we go with a number high enough that we don't expect to hit with fields (maybe `1000`?), we have little risk in breaking that expecation.

**Reply by mattsherman** [reply]
I like the idea of the implicit default being something very high, like `1000`. I think that, along with returning that default when there is no explicit `order` set would solve the issue for most cases.
